pkg/cache/v3/delta.go

```diff
@@ -20,8 +20,16 @@ import (
 	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
 )
 
+type ResourceContainer interface {
+	GetResources(typeURL string) map[string]types.Resource
+
+	GetVersionMap() map[string]map[string]string
+
+	GetVersion(typeURL string) string
+}
```

Comment:
Here is the only tricky thing about this PR. Instead of passing Snapshot to the `respondDelta ` function we now pass the instance of the `ResourceContainer`, which snapshot implements. This allows to easily reuse the function for linear cache. Originally I was planning to make the linear cache to implement the interface, but this doesn't work well (it doesn't make sense for the cache methods to accept typeURL, also GetResources method should return `types.Resource` instead of `map[string]types.Resource`) So I end up adding a wrapper class `cacheWrapper `, which adopts linear cache to the ResourceContainer interface.

Reply:
@s-matyukevich would it be useful to you if I were to add something like:
```go
// List returns a list of resources not indexed by name.
// This is useful when dealing with the linear cache implementation.
func (r Resources) List() []types.Resource {
	resources := []types.Resource{}
	for _, item := range r.Items {
		resources = append(resources, item.Resource)
	}
	return resources
}
```
in the new snapshot API? So now we have a slice when dealing with `types.Resources` instead of having to pass around the map in the linear cache? I think if I understand this right you'd be able to remove the cacheWrapper that way? I guess my goal here is to reduce the amount of boilerplate needed to get the linear cache to work. 

What could we do to the API to remove the need of the `resourceContainer`? If we're going to `1.0` here now is the time to design an API that will stand the test of other cache implementations

Reply:
Adding just this method won't solve the problem as we need also methods to get the version map and system version. Basically, we need something that implements the full interfaces https://github.com/envoyproxy/go-control-plane/pull/459/files#diff-a54688e00c731dac28aa2b2faab04dbd4dffd0aaac139cf0abf98b55efb31963R25-R31